doc/extending/extending_theano_gpu.txt

```diff
@@ -0,0 +1,236 @@
+
+Finally if you require the context before perform, such as during
+make_thunk() to initialize kernels and such, you can access the
+context of your inputs through the type if the variables::
+
+    def make_thunk(self, node, storage_map, compute_map, no_recycling):
+        ctx = node.inputs[0].type.context
+
+Note that GpuArrayType objects also have a `context_name` attribute
+which is the symbolic equivalent of `context`.  It can't be used for
+calls to pygpu or libgpuarray, but it should be used for theano
+operations and variables.
+
+The last place where you might need the context is in the C
+initialization code.  For that you will have to use the :ref:`params
+<extending_op_params>`.  The params type should be
+:class:`theano.gpuarray.type.gpu_context_type` and the params object
+should be a context object from one of your input variables::
+
+    def get_params(self, node):
+        return node.inputs[0].type.context
+
+If you don't have any input variables on the GPU you can follow the
+the example of :class:`theano.gpuarray.basic_ops.GpuFromHost` or
+:class:`theano.gpuarray.basic_ops.GpuEye`.  This is not a case that
+you should encounter often, so it will not be covered further.
+
+Defining New Kernels
+====================
+
+If your op needs to do some transformation on the data, chances are
+that you will need to write a new kernel.  The best way to do this is
+to leverage GpuKernelBase (or CGpuKernelBase if you want to use the
+COp functionality).
+
+For plain GpuKernelBase, you have to define a method called
+gpu_kernels which returns a list of :class:`Kernel
+<theano.gpuarray.basic_ops.Kernel>` objects.  You can define as many
+kernels as you want for a single op.  An example would look like this::
+
+    def gpu_kernels(self, node, name):
+        code = """
+    KERNEL void k(GLOBAL_MEM ga_double *a, ga_size n, ga_size m) {
+        ga_size nb = n < m ? n : m;
+        for (ga_size i = LID_0; i < nb; i += LDIM_0) {
+            a[i*m + i] = 1;
+        }
+    }"""
+        return [Kernel(
+                code=code, name="k",
+                params=[gpuarray.GpuArray, gpuarray.SIZE, gpuarray.SIZE],
+                flags=Kernel.get_flags('float64'))]
+
+If you want to use COp, then you should use ``CGpuKernelBase``
+instead.  It adds a new section to the parsed files whose tag is
+``kernels``.  Inside that section you can define some kernels with
+``#kernel name:params:flags``.
+
+Here ``name`` is the name of the kernel function in the following
+code, ``params`` is a comma-separeted list of numpy typecode names.
+There are three exceptions for ``size_t`` which should be noted as
+``size``, ``ssize_t`` which should be noted as ``ssize`` and a pointer
+which should be noted as ``*``.
+
+``flags`` is a ``|``-separeted list of C kernel flag values (can be
+empty).  The same kernel definition as above would look like this with
+``CGpuKernelBase``::
+
+    #section kernels
+
+    #kernel k : *, size, size : GA_USE_DOUBLE
```

Comment:
We don't need to specify the type that is pointed to, it doesn't matter for the call protocol.

Also, the order of the flags doesn't matter.